omise/omise-magento
# Implementing manual sync for omise payment methods.

#### 1. Objective

Because we can't just only rely on Bank(s)'s response at the time that charge has been made (especially on Internet Banking payment method).
They won't always always return a charge result ('successful', 'failed') status right away after you make a request. Sometimes they take few mins ~ few hours to process a payment before returning a charge result.

Means, if this thing occurs, WooCommerce store has no way to know this result unless we integrate with **[Omise Webhook](https://www.omise.co/api-webhooks)** feature or 'Manual Sync'.

🎦 Added "Sync Order Status" button. Order status is "Payment pending"
<img width="1440" alt="Screen Shot 2020-10-27 at 13 19 56" src="https://user-images.githubusercontent.com/5526195/97264523-b6060000-1857-11eb-84ec-77740ac73ce1.png">


🎦 Changed order status after Manual Sync
<img width="1334" alt="Screen Shot 2020-10-27 at 13 21 38" src="https://user-images.githubusercontent.com/5526195/97264552-c4541c00-1857-11eb-89f2-749a00a966c8.png">


🎦 Updated Invoice after Manual sync
<img width="976" alt="Screen Shot 2020-10-27 at 13 22 30" src="https://user-images.githubusercontent.com/5526195/97264579-d1710b00-1857-11eb-8ea8-4df7c87ca4ad.png">


🎦 Updated order comment after manual sync
<img width="1324" alt="Screen Shot 2020-10-27 at 13 25 40" src="https://user-images.githubusercontent.com/5526195/97264689-02514000-1858-11eb-8a01-b6fa67523be0.png">


Internal Ticket: https://omise.atlassian.net/browse/FES-22

#### 2. Description of change

- Created new plugin class _Plugin/BtnOrderViewPlugin.php_ to add button "Sync Order Status".
- Created controller class _Controller/Adminhtml/Ordersync/Index.php_ for new backend action.
- Registered new controller class in _etc/adminhtml/routes.xml_
- Registered new plugin class in  _etc/adminhtml/di.xml_
- Modifed helper class to add new library functions _Helper/OmiseHelper.php_
- Create new php class Model/SyncStatus.php which checks omise charge status and 

#### 3. Quality assurance

**🔧 Environments:**

- **Platform version**: Magento CE 2.4.
- **Omise plugin version**: Omise-Magento 2.13.
- **PHP version**: 7.4.1.

**✏️ Details:**

- Need to make sure that webhook isn't configured in Omise dashboard.

**Steps for testing**
1. Place an order using omise method method. e.g. Promptpay.
2. Go to Admin dashboard > Orders > View Order
3. Click on the button to "Sync Order Status"
4. Order status should update with Omise charge. Order comment should get added to order as well.


#### 4. Impact of the change

Order status should get updated using "Order Sync Status" button.

#### 5. Priority of change

Normal, High or Immediate.

#### 6. Additional Notes

Any further information that you would like to add.
